synapse/handlers/oidc.py

```diff
@@ -1035,6 +1139,161 @@ def _remote_id_from_userinfo(self, userinfo: UserInfo) -> str:
+        #   Tokens. If the Logout Token is encrypted, it SHOULD replicate the
+        #   iss (issuer) claim in the JWT Header Parameters, as specified in
+        #   Section 5.3 of [JWT].
+        alg_values = metadata.get("id_token_signing_alg_values_supported", ["RS256"])
+        jwt = JsonWebToken(alg_values)
+
+        # As per sec. 2.6:
+        #    3. Validate the iss, aud, and iat Claims in the same way they are
+        #       validated in ID Tokens.
+        # Which means the audience should contain Synapse's client_id and the
+        # issuer should be the IdP issuer
+        claim_options = {
+            "iss": {"values": [self.issuer]},
+            "aud": {"values": [self.client_id]},
+        }
+
+        logger.debug("Attempting to decode JWT logout_token %r", logout_token)
+
+        # Try to decode the keys in cache first, then retry by forcing the keys
+        # to be reloaded
+        jwk_set = await self.load_jwks()
+        try:
+            claims = jwt.decode(
+                logout_token,
+                key=jwk_set,
+                claims_cls=LogoutToken,
+                claims_options=claim_options,
+            )
+        except ValueError:
+            logger.info("Reloading JWKS after decode error")
+            jwk_set = await self.load_jwks(force=True)  # try reloading the jwks
+            claims = jwt.decode(
+                logout_token,
+                key=jwk_set,
+                claims_cls=LogoutToken,
+                claims_options=claim_options,
+            )
+
+        logger.debug("Decoded logout_token JWT %r; validating", claims)
+        claims.validate(leeway=120)  # allows 2 min of clock skew
+
+        # As per sec. 2.6:
+        #    4. Verify that the Logout Token contains a sub Claim, a sid Claim,
+        #       or both.
+        #    5. Verify that the Logout Token contains an events Claim whose
+        #       value is JSON object containing the member name
+        #       http://schemas.openid.net/event/backchannel-logout.
+        #    6. Verify that the Logout Token does not contain a nonce Claim.
+        # This is all verified by the LogoutToken claims class, so at this
+        # point the `sid` claim exists and is a string.
+        sid: str = claims.get("sid")
+
+        # Find in the store if we have a device associated with that SID
+        devices = await self._store.get_devices_by_auth_provider_session_id(
+            auth_provider_id=self.idp_id,
+            auth_provider_session_id=sid,
+        )
+        for device in devices:
+            user_id = device["user_id"]
+            device_id = device["device_id"]
+            logger.info(
+                "Logging out %r (device %r) via OIDC backchannel logout (sid %r).",
+                user_id,
+                device_id,
+                sid,
+            )
+            await self._device_handler.delete_device(user_id, device_id)
```

Comment:
Probably better to use the `delete_devices` method (it looks like most of the code is the same...feels like something we should clean-up).

Reply:
There is a theoretical code path here where `get_devices_by_auth_provider_session_id` returns devices from two different users. That would indicate a faulty/weird OIDC Provider or a weird user mapping provider, but still possible. `delete_devices` assumes it's the same `user_id` for each device

Reply:
Can we add a comment about this?